docs/guides/styling.md

```diff
@@ -766,6 +766,76 @@ Other CSS-in-JS libraries will have a similar setup. If you've got a CSS framewo
 
 NOTE: You may run into hydration warnings when using Styled Components. Hopefully [this issue][styled-components-issue] will be fixed soon.
 
+## CSS Modules
+
+<docs-warning>This feature is unstable and currently only available behind a feature flag. We're confident in the use cases it solves but the API and implementation may change in the future.</docs-warning>
+
+<docs-warning>Since all CSS Modules styles end up in a single CSS file at the end of the build, we recommend that you only use this styling approach for smaller applications, or sparingly within larger applications. If you're concerned about bundle size, you should probably look at [Tailwind](#tailwind-css) or [Regular Stylesheets](#regular-stylesheets) instead.
+</docs-warning>
+
+To enable [CSS Modules], set the `unstable_cssBundle` feature flag in `remix.config.js`.
+
+```js filename=remix.config.js
+/** @type {import('@remix-run/dev').AppConfig} */
+module.exports = {
+  future: {
+    unstable_cssBundle: true,
+  },
+  // ...
+};
+```
+
+With this feature flag enabled, the Remix compiler will now generate a single CSS file containing all CSS Modules styles in your application. Note that any [regular stylesheet imports](http://localhost:3000/docs/en/v1/guides/styling#regular-stylesheets) will remain as separate files.
+
+Unlike many other tools in the React ecosystem, insertion of this CSS bundle into the page is not handled for you automatically. Instead, we ensure that you always have control over the link tags on your page. This lets you decide where the CSS bundle is inserted and in what order relative to other style sheets.
+
+To get access to the CSS bundle, first install the `@remix-run/css-bundle` package.
+
+```sh
+npm install @remix-run/css-bundle
+```
+
+Then, import the CSS bundle's `href` and add it to a link descriptor, most likely in `root.tsx` so that it applies to your entire application.
+
+```tsx filename=root.tsx lines=[2,6]
+import type { LinksFunction } from "@remix-run/node"; // or cloudflare/deno
+import cssBundleHref from "@remix-run/css-bundle";
```

Comment:
I'm wondering if we shouldn't use a more specific virtual here for CSS Modules. If we take a similar approach with raw CSS imports and Vanilla Extract, it could be nice to have access to each of those bundles independently (ex: your admin routes are styled w/ CSS Modules, marketing pages use some other bundle)

Reply:
Perhaps we could achieve both?

```ts
// all bundles in one stylesheet
import cssBundleStylesheet from "@remix-run/css-bundle";

// split up
import cssModulesStylesheet from "@remix-run/css-bundle/css-modules";
import vanillaExtractStylesheet from "@remix-run/css-bundle/vanilla-extract";
```

Reply:
We don't get that level of granularity since we're now hooking into esbuild's CSS loader, which means we'll only get a single file at the end. I'm not convinced this is a big enough use case to warrant implementing our own CSS bundling logic, compared to bundle splitting of CSS for example.

It also sounds like a bit of a fragile pattern to me since nothing enforces that you're only using CSS Modules / Vanilla Extract / etc. within a given section of your application, leading to missing styles if you happen to use these features elsewhere.

Reply:
Good points. Perhaps another thing we attempt to solve later by splitting CSS at the route level.